muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/communicator/rest/CommunicatorRESTService.java

```diff
@@ -1,1064 +1,1078 @@
+        }
+      } else {
+        // Trying to feed group ids when you don't have permission greets you with bad request
+        return Response.status(Status.BAD_REQUEST).build();
+      }
+    }
+
+    // Workspace members
+
+    if (!CollectionUtils.isEmpty(newMessage.getRecipientStudentsWorkspaceIds())) {
+      workspaceStudentRecipients = new ArrayList<WorkspaceEntity>();
+      
+      for (Long workspaceId : newMessage.getRecipientStudentsWorkspaceIds()) {
+        WorkspaceEntity workspaceEntity = workspaceEntityController.findWorkspaceEntityById(workspaceId);
+  
+        if (sessionController.hasPermission(CommunicatorPermissionCollection.COMMUNICATOR_WORKSPACE_MESSAGING, workspaceEntity))
+          workspaceStudentRecipients.add(workspaceEntity);
+        else
+          return Response.status(Status.BAD_REQUEST).build();
+      }
+    }
+
+    if (!CollectionUtils.isEmpty(newMessage.getRecipientTeachersWorkspaceIds())) {
+      workspaceTeacherRecipients = new ArrayList<WorkspaceEntity>();
+      
+      for (Long workspaceId : newMessage.getRecipientTeachersWorkspaceIds()) {
+        WorkspaceEntity workspaceEntity = workspaceEntityController.findWorkspaceEntityById(workspaceId);
+  
+        if (sessionController.hasPermission(CommunicatorPermissionCollection.COMMUNICATOR_WORKSPACE_MESSAGING, workspaceEntity))
+          workspaceTeacherRecipients.add(workspaceEntity);
+        else
+          return Response.status(Status.BAD_REQUEST).build();
+      }
+    }
+    
+    if (StringUtils.isBlank(newMessage.getCategoryName())) {
+      return Response.status(Status.BAD_REQUEST).entity("CategoryName missing").build();
+    }
+
+    // TODO Category not existing at this point would technically indicate an invalid state
+    CommunicatorMessageCategory categoryEntity = communicatorController.persistCategory(newMessage.getCategoryName());
+    
+    CommunicatorMessage message = communicatorController.createMessage(communicatorMessageId, userEntity, 
+        recipients, userGroupRecipients, workspaceStudentRecipients, workspaceTeacherRecipients, categoryEntity, 
+        newMessage.getCaption(), newMessage.getContent(), tagList);
+    
+    sendNewMessageNotifications(message);
+    
+    return Response.ok(
+      restModels.restFullMessage(message)
+    ).build();
+  }
+
+  private void sendNewMessageNotifications(CommunicatorMessage message) {
+    List<CommunicatorMessageRecipient> recipients = communicatorController.listAllCommunicatorMessageRecipients(message);
+    
+    for (CommunicatorMessageRecipient recipient : recipients) {
+      // Don't notify the sender in case he sent message to himself
+      
+      if (recipient.getRecipient() != message.getSender()) {
+        communicatorMessageSentEvent.fire(new CommunicatorMessageSent(message.getId(), recipient.getRecipient(), baseUrl));
+        
+        UserEntity recipientEntity = userEntityController.findUserEntityById(recipient.getRecipient());
```

Comment:
Do this only if the recipient is not part of a group receiving the message. (`recipient` should have the group information so _maybe_ just by checking if that is null should work.)

Reply:
The check is correct but needs to be in its own if.